[16364] Switch WorkPackage show view over to AngularJS (aka fullscreen view)

timhabermaas opened the pull request:
https://community.openproject.org/work_packages/16364
- [ ] fix specs
- [ ] clean up directives/code duplication
- [ ] being in split view and clicking on a work package does not update the right pane
- [x] "Watchers" does not show watchers and one can't add watchers. see split view for correct implementation
#### Review Feedback

Feedback von @lindenthal und @basiszwo
- [x] Inline editing of WP title missing
- [x] project context is being lost when switching to fullscreen view `/project/seeded_project/work_packages/1` vs `/work_packages/1`
- [x] "more" Button with actions in dropdown is missing [1]
- [x] "Edit" Button should link to old edit view (same as in split view) [1]
- [x] cog icon on top right does not make sense in fullscreen view as it provides list settings
  - replace with "more" Button?
- [x] missing full screen icon (already included in icon set)
- [x] scrollbar should not be visible when there's nothing to scroll (@rebellium)
- [x] Validate/Fix the actions which should still work (e.g. watching a work package); see OpenProject's work package
- [x] Title should be "Bug #123: WP TITLE" where "WP title is editable" and "Bug" is not a link
- [x] carry query path of url to fullscreen view url so we are properly being redirected when switching back to list or split view
- [x] dropdown actions in "new work package" button are missing so it is not possible to create work packages from the full screen view. (should redirect to create mode in split view)

[1] we initially place the buttons in the top right area. Later @rebellium needs to move them to a bottom bar.
#### Found issues
- [ ] [breadcrumb does nto show project](https://github.com/opf/openproject/pull/3366#issuecomment-140994275)
- [ ] [Can't change view on small resolutions](https://github.com/opf/openproject/pull/3366#issuecomment-140995420)
- [ ] [Request-loop for non-existing work packages](https://github.com/opf/openproject/pull/3366#issuecomment-140998345)
- [ ] [toolbar line-breaking](https://github.com/opf/openproject/pull/3366#issuecomment-141000902)
- [ ] [rendering of anchors](https://github.com/opf/openproject/pull/3366#issuecomment-141015332)
#### Questions
- is it possible to provide a full edit view? how much work would this cause? maybe also in split view mode?
- what happens if someone has no right to view the work package?
#### Out of scope

Create new pull requests for these bugs
- [x] Confirm when deleting an image (#3437)
- [x] Upload image then try to edit the description => error 'conflicting modifications' (#3436)
#### Won't apply
- [x] "new work package" Link in sidebar is linked to the old new view. We should relink it to the new create view (see https://github.com/opf/openproject/pull/3366#issuecomment-135332990)

timhabermaas committed:
Add fullscreen view for WorkPackage.

timhabermaas committed:
Point work package link from table to new angular show view.

timhabermaas committed:
Fix listing of attributes on fullscreen WorkPackage page.

timhabermaas committed:
Fix buttons for switching between views on WorkPackage show page.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://timhabermaas-16364-fullscreen-work-package-view.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://timhabermaas-16364-fullscreen-work-package-view.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

timhabermaas committed:
Fix non-camelcased name.

Thanks, @houndci.

ulferts commented:
Thanks for the PR, @timhabermaas. It feels good to finally see it all fit together. Let me know if it is too early to give feedback. 

Here is what I noticed when skimming over the functionality:

The url seems wrong to me:
![image](https://cloud.githubusercontent.com/assets/617519/9300246/18efcfba-44c4-11e5-8bb2-2b8b1c5d681d.png)
I believe it should be `/work_packages/:id` (please see the work package as well). That should also render the problem of having to fix the routes in the rails world non existent. However, it might introduce new questions when switching between full screen view and details pane/table view especially when having started from the full screen view.

The layout needs some adjustments (especially on large displays:
![image](https://cloud.githubusercontent.com/assets/617519/9300191/a980f744-44c3-11e5-8167-5b226d7de4c8.png)

There is no scrollbar on the page leading to content becoming impossible to reach:
![image](https://cloud.githubusercontent.com/assets/617519/9300203/d19275aa-44c3-11e5-9b66-b9088fa0db68.png)
This is a behaviour inherited from the list view.

The width of the overview content changes depending on whether the "show all" button is enabled or not:
![image](https://cloud.githubusercontent.com/assets/617519/9300215/ee9f3516-44c3-11e5-9fb8-9e05b20ec768.png)
vs.
![image](https://cloud.githubusercontent.com/assets/617519/9300223/f99dfefc-44c3-11e5-99de-8510673f359b.png)

The "Relations" tab is not styled properly (the other's look unchanged):
![image](https://cloud.githubusercontent.com/assets/617519/9300285/8b4f4586-44c4-11e5-97e6-8ad3fec70157.png)

It's not possible to create a work package via the view:
![image](https://cloud.githubusercontent.com/assets/617519/9300304/a91ff9d4-44c4-11e5-903f-41759906de20.png)

The settings button is non functional:
![image](https://cloud.githubusercontent.com/assets/617519/9300337/0509416a-44c5-11e5-98d4-ede9f765de2f.png)

There are no tests.

timhabermaas committed:
Move Angular's WorkPackage show view to /work_packages/:id.

timhabermaas commented:
Thanks for the feedback, @ulferts. It's kinda early, but still appreciated! I wanted to add the `work-in-progress` label to this PR, but ~~don't~~ didn't seem to have the rights to do so. ~~Could someone add it, please?~~ (thanks @NobodysNightmare)

> The url seems wrong to me:

I plan to change it, but need to move some code around first (abstract parent controller depends on the URL being `/projects/:foo/...`).

> That should also render the problem of having to fix the routes in the rails world non existent.

Except for `/work_packages/:id#comment_2`. Anchors will probably break when rendering the page client side.

I'm aware of the design glitches and will fix them at the end. Your tip "This is a behaviour inherited from the list view." is useful, thanks!

> There are no tests.

What kind of tests would you like to see? E.g. there seem to be integration tests using RSpec and some JS integration tests for the work packages index page. Which one should I prefer to modify/extend?

timhabermaas commented on frontend/app/routing.js:
```diff
@@ -67,8 +67,51 @@ angular.module('openproject')
+
     .state('work-packages.list', {
-      url: '',
+      url: '/projects/:projectIdentifier/work_packages?query_id',
```
@floriank How bad of an idea is this? Are there other routes besides `/projects/:projectIdentifier/` where the work packages can be nested under?

floriank replied:
For starters, there is a whole route dedicated to work_packages without a project scope: 

```
/work_packages
```

will show the complete list independent of a project. I will look into this a bit deeper tomorrow.

also, is the `query_id` part necessary? Arbitrary params should be accessible via `$params` or `$stateParams` (I mix those up all the time)

NobodysNightmare replied:
I can answer to this subpart of the question

> Are there other routes besides /projects/:projectIdentifier/ where the work packages can be nested under?

Nope. We only nest work packages under projects or don't nest them at all. My fantasy also can't come up with other useful nestings, so we should be quite safe from this side ;)

timhabermaas committed:
Fix switching back to list view from fullscreen view.

timhabermaas committed:
Fix route in WorkPackagesPage; fixes some specs.

rebellium committed:
Fix split-view styling issues

timhabermaas committed:
Fix undefined errors on show page.

timhabermaas committed:
Fix markup.

timhabermaas committed:
Redirect user to default tab when landing on work package show page.

timhabermaas committed:
Add colon after work package subject in header.

timhabermaas committed:
Change layout of workpackage title on show view.

timhabermaas commented:
@lindenthal @basiszwo Thanks for the feedback (would've preferred it in a different comment instead of editing the first post, though). I'm working through the list right now.

Both of these issues are already present on `dev`:

> Confirm when deleting an image
> Upload image then try to edit the description => error 'conflicting modifications'

I'd prefer to try to fix them in a different branch. Agreed?

basiszwo commented:
agree

timhabermaas committed:
Preserve query parameters when switching to fullscreen view.

timhabermaas committed:
Make creating work packages from show view work.

ulferts commented on frontend/app/routing.js:
```diff
@@ -79,7 +133,7 @@ angular.module('openproject')
     })
     .state('work-packages.list.details', {
-      url: '/{workPackageId:[0-9]+}?query_props',
+      url: '/details/{workPackageId:[0-9]+}?query_props',
```
This change will break existing links people might already have stored outside of the system, .e.g by sending it via mail. Why is it necessary? The former route seemed to be uniq to me as well.

timhabermaas replied:
It's to differentiate between the fullscreen show view and the split view. If I don't change one of them both have `/work_packages/:id/:tab` as the URL. Changing the show view URL seemed more cumbersome to me.

Old links should still work, but they end up pointing to the fullscreen view instead of the split view now. IMO no big deal: Same information, different layout.

ulferts replied:
I think it is not working as you intended, then. 
In the dev and release branches, this is the url for opening a details pane (table + details pane opened with the overview tab): `/projects/:project_id/work_packages/:id/overview`. In this branch, it will yield an empty page. Instead one has to query with `/projects/:project_id/work_packages/details/:id/overview`.

When simply removing the `details` part everything seems to be working again. I am not an expert on the routing but I believe that the state is scoped somehow to only be available scoped under '/projects/:project_id'.

timhabermaas replied:
Ah yeah, missed the nested routes for the show page. I'll fix them once I managed to get the drop-down menu working on the show page.

timhabermaas committed:
Fix work package's show and create routes which were shadowed.

timhabermaas committed:
Add edit button to Angular show view.

basiszwo commented:
@timhabermaas I updated the description (reordered the items)

basiszwo commented:
to get the fullscreen icon you need to set the `icon-fullscreen` css class

timhabermaas committed:
Use correct icon for fullscreen button.

timhabermaas commented:
>  "new work package" Link in sidebar is linked to the old new view. We should relink it to the new create view

Which isn't as trivial as thought: We need to provide a default work package type when linking to the angular new form since there's no option to select the type (URL looks like `/projects/:project_id/work_packages/create_new?type=1`).

basiszwo commented:
as it is not easy to link to a default type and change it in the form afterwards we are removing the "new" action from the sidebar. This was already in discussion with @lindenthal and now it's going to happen :smile:

NobodysNightmare commented:
> We need to provide a default work package type when linking to the angular new form

Not sure if this hint applies, but the APIv3 will assume a default type already when a client does not indicate one. E.g. `POST /api/v3/projects/1/work_packages/form` with an empty payload _should_ return a form, where the payload has the default type set.

What I do not know is how/if this applies to the frontend problems you have, but it might be helpful to know.

timhabermaas committed:
Make create_new work package view work when no type is set.

timhabermaas commented:
@NobodysNightmare Thanks! I didn't know this. This made the fix pretty easy: 726051efc8477d92b57cf5f59731bf077a2ec035

basiszwo commented:
see http://clients.zzmedia.net/openproject/workpackages/fullscreen/small/ for an arrangement of the top right icons. The settings (cog) icon on the most upper right will contain some links to the administration area of openproject. The links show in list view don't make sense.
We come up with links soon.

basiszwo commented:
@timhabermaas the correct icons we need to use are not yet included. Look at https://github.com/opf/openproject/pull/3424/files

we need to replace all 3 icons
- icon-fullscreen-view
- icon-list-view
- icon-split-view

timhabermaas committed:
Add watch icon to angular work package show view.

timhabermaas committed:
Order toolbar buttons according to screen design.

timhabermaas committed:
Add missing semicolons.

ulferts commented:
@timhabermaas can you please merge/rebase as it is getting harder and harder to compare and work with this branch. e.g. there are some migrations in dev now that will cause breakages when switching to your branch.

timhabermaas commented:
@ulferts I rebased.

timhabermaas committed:
Use new icons for split view toolbar.

HDinger committed:
resized input fields so that they are on full width

HDinger commented:
I started with the styling of the input fields in https://github.com/timhabermaas/openproject/pull/2.

@timhabermaas @basiszwo @lindenthal : could you please have a quick look at the new styling?

HDinger committed:
set overflow value to auto

HDinger commented:
This should fix the problem with the scrollbar: https://github.com/timhabermaas/openproject/pull/3

HDinger committed:
inserted overflow-values

HDinger commented:
This PR deals with text-overflow values (needed for example for headlines or comments with really long words): https://github.com/timhabermaas/openproject/pull/4

timhabermaas committed:
Make actions menu work on show page.

timhabermaas committed:
Make @houndci happy.

timhabermaas committed:
Remove link to new work package from sidebar menu.

timhabermaas committed:
Attempt to make inline editing of work package subject work

timhabermaas committed:
Fix broken /projects/:id/work_packages/:id URL.

This also fixes some inconsistencies with sidebar rendering when
reloading the page.

basiszwo committed:
Tmp fix css for subject inplace edit to display type and subject inline

HDinger commented:
This PR resizes the headlines to full width: https://github.com/timhabermaas/openproject/pull/6

@basiszwo : can you please have a look at it?

timhabermaas committed:
Fix links from work packages table to fullscreen view.

timhabermaas committed:
Only display dropzone if "Show all" is selected.

basiszwo committed:
Print work package tabbar with capitals

basiszwo committed:
Align left panel content with work package headline

basiszwo committed:
Increase line height for wp title in split view

basiszwo committed:
Decrease padding for activity lists

basiszwo committed:
Decrease font-size for date in activity list

basiszwo committed:
Decrease font-size for icon in relations tab

HDinger committed:
This commit avoids horizontal scrollbars when there is nothing to scroll.

timhabermaas committed:
Hide watch/unwatch button if user has no rights to use it.

timhabermaas committed:
Remove unused variable.

timhabermaas committed:
Only display "more" drop-down if user has proper rights.

timhabermaas committed:
Fix routing errors.

timhabermaas committed:
Add missing redirect rule for work_packages nested under a project.

timhabermaas committed:
Remove obsolete css rule.

timhabermaas committed:
Fix routing specs.

timhabermaas committed:
[WIP] fix attachment spec

TODO: figure out why Firefox wants to download attachment instead of
opening it.

ulferts committed:
disable reloadOnSearch for work packages list

This prevents reinstantiating a controller on changes to the query
params of a url.

In this case, we can now use $state.go instead of working around it via
$location.

We still have to prevent the suburi from being included twice, though.
Because $state (and $location) for that matter will append the sub-uri
anyway, we can't add it ourselves or we will end up with it being
prepended twice.

ulferts committed:
fix styling

ulferts committed:
use PathHelper to avoid code duplication

ulferts committed:
remove base uri from project path calculation

The base uri is already factored in by the ui-router

ulferts committed:
rely on $location instead of $state for query params

ulferts commented:
I addressed the sub-uri problems in https://github.com/timhabermaas/openproject/pull/8

timhabermaas committed:
Hide attachment dropzone if user has no rights to upload files.

timhabermaas committed:
[WIP] [skip ci] fix some karma specs

timhabermaas committed:
[skip ci] [WIP] some cucumber fixes

timhabermaas committed:
Fix missing activities content of work package show page.

timhabermaas committed:
Remove obsolete legacy specs.

timhabermaas committed:
Make removing watchers work on show view.

* Set feature specs concerning watching/unwatching to WIP
* Fix removing watcher spec
* Set adding watcher spec to WIP since it's broken on `dev`

NobodysNightmare commented:
One issue that was brought to me: In the full screen view the first activity is displayed as "updated" (instead of "created") and all fields are marked as changed:

![pasted image at 2015_09_09 11_05 am](https://cloud.githubusercontent.com/assets/453584/9757838/476e8b34-56e3-11e5-9250-d647e3caa516.png)

For the split view it works as expected.

Background: This might be due to the lame way I implemented the recognition of the first activity, which was later altered even more by @oliverguenther .

**tl; dr:** You have to tell a user activity whether it `isInitial` explicitly. I am sorry for that. When I first did it, it looked like a quick and "good enough" solution :/

basiszwo commented:
- "watch" icon shows `true` when watching a WP
- "Watchers" does not show watchers and one can't add watchers. see split view for correct implementation

![bildschirmfoto 2015-09-09 um 11 53 20](https://cloud.githubusercontent.com/assets/24071/9758857/64926a86-56e9-11e5-875f-646a5fec62e2.PNG)

timhabermaas committed:
Fix broken indicator of initial activity on work package show page.

timhabermaas committed:
Move isInitialActivity to ActivityService.

timhabermaas committed:
Remove unused function.

timhabermaas commented:
@NobodysNightmare This should be fixed with: 3f5256c9069e1548eff3b7f59720e31359ed11e4 (and 5b3779147545b789c1e602f26dcbce1e1a9abbc7). The `isInitialActivity` wasn't present at all. I love AngularJS for not complaining about such obvious errors. :angry:

timhabermaas commented:
There's another issue with the split view: Double clicking on an entry doesn't change the shown work package. Removing `reloadOnSearch` in the router fixes this, but also causes the work package list to be reloaded.

basiszwo committed:
Adjust specs and cukes to new fullscreen view

timhabermaas committed:
Fix karma specs.

basiszwo committed:
Provide bulk destroy functionality for work package

This is considered to be hacky and needs to be removed soonest.

basiszwo committed:
Split view panel is always visible

Before the split view panel was hidden when screensize was < 1280px.
This is no longer the case as we need it to create and edit the work
packages.

timhabermaas committed:
[WIP] Fix protractor specs.

timhabermaas committed:
Remove reloadOnSearch: false flags from detail tabs.

Switching between different details views was prevented by that flag.

timhabermaas committed:
Fix attachments directive test after rebase.

Stub the `links` attribute of a work package which is usually present
when using the API.

basiszwo committed:
Assert redirect for atom representation of work package

basiszwo committed:
Attachments are no longer being rendered inline

basiszwo committed:
Adjust accessibility spec

We currently don't have a full edit view.

basiszwo committed:
Skip work package show view accessibility specs

basiszwo committed:
Adjust diff on show cuke

Need to reorganize markup to properly identify selectors for testing.

basiszwo committed:
Merge pull request #9 from basiszwo/16364-fullscreen-work-package-view

[16364] Adjust specs and cukes to new fullscreen view

NobodysNightmare commented:
### Breadcrumb does not shows work packages project

#### Reproduction
- open a work package using a direct link (with a project in the URL)
  - e.g. https://community.openproject.org/work_packages/21497

#### Expected (previous behaviour)
- The breadcrumb shows the work packages project
- there is a project specific side menu

![image](https://cloud.githubusercontent.com/assets/453584/9927603/5be0bde2-5d20-11e5-8c6e-1290b10a7fe1.png)

#### Actual
- breadcrumb shows no project
- no side menu

![image](https://cloud.githubusercontent.com/assets/453584/9927653/ba5635d2-5d20-11e5-9f85-f8eaad34d232.png)

NobodysNightmare commented:
### Can't change view on small resolutions

#### Reproduction
- use a viewport of 1280px or less width
- open the WP list
- click on a row

#### Expected behaviour
- no inapropriate view mode for my resolution is opened
- in the top right I can choose from available view modes for my resolution
  - list view (always available)
  - split view (IIRC, only available above the break point of 1280px width)
  - fullscreen view (always available)

![image](https://cloud.githubusercontent.com/assets/453584/9927866/439b8eae-5d22-11e5-8fc3-c5ef5347627b.png)

#### Actual
- split view opens
- I have no option to change the view mode

![image](https://cloud.githubusercontent.com/assets/453584/9927841/1e8bfd1a-5d22-11e5-84aa-f4634e4bf4c4.png)

NobodysNightmare commented:
### Request-loop on non-existing work packages

#### Reproduction
- Open the URL for a non-existing (e.g. deleted) work package
  - e.g. community.openproject.org/work_packages/99999 (be careful: better try your local instance!)
- important: do not append the `/activity` part

#### Expected
- some kind of error page, indicating that this is not the work package I am looking for

##### Actual
- a wild request-loop appears, causing some 404 flooding:

![image](https://cloud.githubusercontent.com/assets/453584/9928012/20dd0d92-5d23-11e5-9ef8-8c81395c71e6.png)

I guess this has something to do with `work_packages/:id` vs `work_packages/:id/activity`, since the `activity` part in my adress bar began flickering hard...

#### Notes

This also works for work packages I am not allowed to see, but that actually would exist...

NobodysNightmare commented:
### Subject and toolbar do not always line-break correctly

#### Reproduction
- use a long subject (multiple words)
- look at the fullscreen view at different widths

#### Expected
- line breaking always occurs in a _good looking_™ manner

#### Actual

At different widths the line breaking is sometimes good, sometimes not so...

good:
![break-1](https://cloud.githubusercontent.com/assets/453584/9928285/6e88c048-5d25-11e5-9a3e-4a06df109972.PNG)

not ideal (space on top is enough):
![break-2](https://cloud.githubusercontent.com/assets/453584/9928286/6e896ade-5d25-11e5-8943-c80837af75d0.PNG)

good (IMO):
![break-3](https://cloud.githubusercontent.com/assets/453584/9928287/6e89c9e8-5d25-11e5-84b9-f8fe87706b17.PNG)

bad (hides part of the subject):
![break-4](https://cloud.githubusercontent.com/assets/453584/9928284/6e874fe2-5d25-11e5-83b9-7b3e042b61c3.PNG)

jonasheinrich commented:
### Wrong Rendering of anchor

#### Reproduction

Create a work package with a description containing an h1-element.

#### Actual

The anchor is one line too low and thus, under the text of the description. Additionally, it is always displayed.

#### Expected

The anchor is in the correct line. Furthermore, the anchor is only visible when you hover over it. 

<img width="242" alt="bildschirmfoto 2015-09-17 um 10 29 25" src="https://cloud.githubusercontent.com/assets/5173966/9929079/cf670b9a-5d2a-11e5-9b23-cf401b7e9815.png">
